### PR TITLE
helper-cli: Add a command to download results from Postgres

### PR DESCRIPTION
**Pull request by fviernau**
The `cli` module provides the UploadResultToPostgresCommand. Add a
command to retrieve the previously uploaded commands. This can be useful
to synchronize a local directory with the database incrementally, in
order to play around with the latest data on the local machine.

Use md5sums to detect broken files, so that the download is gracefully
resumed after a hard exit or a connectivity issue. The need for md5sums
prevented the re-use of the LocalFileStorage class.

### CODE REVIEW

File: helper-cli/src/main/kotlin/commands/DownloadResultsFromPostgresCommand.kt [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        // the gzip extension can help speed this up https://github.com/pramsey/pgsql-gzip.
+
+        val query = """
+        SELECT 
+            t.$columnName::TEXT AS result
+        FROM 
+            $tableName t 
+        WHERE
+            id = ?;
+        """.trimMargin()
```
Comment (sschuberth) [inline helper-cli/src/main/kotlin/commands/DownloadResultsFromPostgresCommand.kt:126]: You're not using any margin marker, so should this be `trimIndent` instead? Also, the query string (lines 120 to 125) should probably be indented.

File: helper-cli/src/main/kotlin/commands/DownloadResultsFromPostgresCommand.kt [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+private class OrtResultStorage(private val storageDir: File) {
+    fun getIds(): Set<Int> {
+        val ids = storageDir.listFiles { file, _ -> file.isDirectory }.mapNotNull {
+            try {
+                it.name.toInt()
+            } catch (e: NumberFormatException) {
+                null
+            }
+        }
```
Comment (sschuberth) [inline helper-cli/src/main/kotlin/commands/DownloadResultsFromPostgresCommand.kt:176]: Maybe better use `runCatching` / `getOrNull`?

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: helper-cli/src/main/kotlin/commands/DownloadResultsFromPostgresCommand.kt (hunk 1 above)
Comment (fviernau) [reply helper-cli/src/main/kotlin/commands/DownloadResultsFromPostgresCommand.kt]: Just noticed that it fits in one line...

File: helper-cli/src/main/kotlin/commands/DownloadResultsFromPostgresCommand.kt [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+            resultSet.getString("result")
+        } else {
+            null
+        }
+    }
+
+    private fun fetchRowIds(connection: Connection): Set<Int> {
+        val query = """
+            SELECT t.id FROM $tableName t WHERE id >= ?;
+        """.trimIndent()
```
Comment (sschuberth) [inline helper-cli/src/main/kotlin/commands/DownloadResultsFromPostgresCommand.kt:133]: I guess this then also fits into a single line...

File: helper-cli/src/main/kotlin/commands/DownloadResultsFromPostgresCommand.kt (hunk 2 above)
Comment (mnonnenmacher) [reply helper-cli/src/main/kotlin/commands/DownloadResultsFromPostgresCommand.kt]: Or even better use `toIntOrNull()`.

File: helper-cli/src/main/kotlin/commands/DownloadResultsFromPostgresCommand.kt [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    override fun run() {
+        if (!outputDir.exists()) outputDir.mkdirs()
+
+        val localStorage = OrtResultStorage(outputDir)
+        val localRowIds = localStorage.getIds()
+
+        openDatabaseConnection().use { connection ->
+            val remoteRowIds = fetchRowIds(connection)
+            val missingRowIds = remoteRowIds - localRowIds
+            println("Requested the download of ${remoteRowIds.size} of which ${missingRowIds.size} are missing.")
```
Comment (mnonnenmacher) [inline helper-cli/src/main/kotlin/commands/DownloadResultsFromPostgresCommand.kt:99]: "results" or "rows" is missing after `${remoteRowIds.size}`.

File: helper-cli/src/main/kotlin/commands/DownloadResultsFromPostgresCommand.kt [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+            missingRowIds.forEachIndexed { i, rowId ->
+                println("[${i + 1} / ${missingRowIds.size}] downloading row $rowId")
+
+                val duration = measureTime {
+                    val ortResultJson = fetchOrtResult(connection, rowId)
+
+                    if (ortResultJson != null) localStorage.add(rowId, ortResultJson)
+                }
+
+                println("[${i + 1} / ${missingRowIds.size}] took ${duration.toString(DurationUnit.SECONDS)}s.")
```
Comment (mnonnenmacher) [inline helper-cli/src/main/kotlin/commands/DownloadResultsFromPostgresCommand.kt:110]: If there is no requirement to always print the duration in seconds you could use the default `toString()` which chooses matching units automatically.

File: helper-cli/src/main/kotlin/commands/DownloadResultsFromPostgresCommand.kt [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+            }
+        }
+    }
+
+    private fun fetchOrtResult(connection: Connection, rowId: Int): String? {
+        // TODO: The transmission is relatively slow currently due to lack of compression. Check whether
+        // the gzip extension can help speed this up https://github.com/pramsey/pgsql-gzip.
+
+        val query = "SELECT t.$columnName::TEXT AS result FROM $tableName t WHERE id = ?;"
```
Comment (mnonnenmacher) [inline helper-cli/src/main/kotlin/commands/DownloadResultsFromPostgresCommand.kt:119]: If you would reuse the `OrtResults` table from the upload command you could get rid of the SQL related boilerplate code.

File: helper-cli/src/main/kotlin/commands/DownloadResultsFromPostgresCommand.kt [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    }
+
+    private fun openDatabaseConnection(): Connection {
+        val storageConfig = OrtConfiguration.load(configArguments, configFile).scanner.storages?.values
+            ?.filterIsInstance<PostgresStorageConfiguration>()?.firstOrNull()
+            ?: throw IllegalArgumentException("postgresStorage not configured.")
+
+        return DatabaseUtils.createHikariDataSource(
+            config = storageConfig,
+            applicationNameSuffix = ORTH_NAME,
```
Comment (mnonnenmacher) [inline helper-cli/src/main/kotlin/commands/DownloadResultsFromPostgresCommand.kt:153]: You could make a separate commit and align this in the upload command.

Comment (fviernau) [reply helper-cli/src/main/kotlin/commands/DownloadResultsFromPostgresCommand.kt:153]: The upload command is a sub command of `ort`, while this one is a sub command `orth`.

File: helper-cli/src/main/kotlin/commands/DownloadResultsFromPostgresCommand.kt (hunk 6 above)
Comment (fviernau) [reply helper-cli/src/main/kotlin/commands/DownloadResultsFromPostgresCommand.kt:119]: I'd keep this for now, in particular as I hope to make some PG specific optimizations, see comment above.

**Review by mnonnenmacher — DISMISSED** [review]

**Comment by sschuberth** [comment]
Sorry @fviernau, I clicked on the wrong button and updated this via a merge instead of a rebase. I'll fix this.

**Review by sschuberth — DISMISSED** [review]

**Review by sschuberth — APPROVED** [review]
